fil_fungible_token/src/token/mod.rs

```diff
@@ -80,28 +89,64 @@ where
     }
 }
 
-impl<BS> Token<BS>
+impl<BS, MC> Token<BS, MC>
 where
     BS: IpldStore + Clone,
+    MC: MethodCaller,
 {
     /// Mints the specified value of tokens into an account
     ///
     /// The mint amount must be non-negative or the method returns an error
-    pub fn mint(&mut self, initial_holder: ActorID, value: TokenAmount) -> Result<()> {
+    pub fn mint(
+        &mut self,
+        minter: ActorID,
+        initial_holder: ActorID,
+        value: &TokenAmount,
+        data: &[u8],
+    ) -> Result<()> {
         if value.is_negative() {
             return Err(TokenError::InvalidNegative(format!(
                 "mint amount {} cannot be negative",
                 value
             )));
         }
 
+        let old_state = self.state.clone();
+
         // Increase the balance of the actor and increase total supply
-        self.state.change_balance_by(&self.bs, initial_holder, &value)?;
+        self.transaction(|state, bs| {
+            state.change_balance_by(&bs, initial_holder, value)?;
+            state.change_supply_by(value)?;
+            Ok(())
+        })?;
 
-        // TODO: invoke the receiver hook on the initial_holder
+        // Update state so re-entrant calls see the changes
+        self.flush()?;
```

Comment:
This might flush previously un-flushed state changes. We do need to flush here, but this could surprise someone so it's pointing to something just slightly off in APIs somewhere.

Basically there's a conflict between the need to flush here, and objectives about holding up state changes for batched flush. Maybe something like this is the best compromise available. I can't see how to sensibly avoid a flush between every transfer or mint, even if doing a large batch of them. But then maybe for consistency we give up on the idea of non-persisted changes at the level of Token API methods.